language/move-prover/src/prover_task_runner.rs

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! Prover task runner that runs multiple instances of the prover task and returns
+//! as soon as the fastest instance finishes.
+
+use crate::cli::Options;
+use async_std::task;
+use async_trait::async_trait;
+use futures::{future::FutureExt, pin_mut, select};
+use rand::Rng;
+use std::{
+    process::Output,
+    sync::mpsc::{channel, Sender},
+};
+use tokio::{
+    process::Command,
+    runtime::Runtime,
+    sync::{broadcast, broadcast::Receiver},
+};
+
+#[derive(Debug, Clone)]
+enum BroadcastMsg {
+    Stop,
+}
+
+#[async_trait]
+pub trait ProverTask {
+    type TaskResult: Send + 'static;
+    type TaskId: Send + Copy + 'static;
+
+    /// Initialize the task runner given the number of instances.
+    fn init(&mut self, num_instances: usize) -> Vec<Self::TaskId>;
+
+    /// Run the task with task_id. This function will be called from one of the worker threads.
+    async fn run(&mut self, task_id: Self::TaskId) -> Self::TaskResult;
+}
+
+pub struct ProverTaskRunner();
+
+impl ProverTaskRunner {
+    /// Run `num_instances` instances of the prover `task` and returns the task id
+    /// as well as the result of the fastest running instance.
+    pub fn run_tasks<T>(mut task: T, num_instances: usize) -> (T::TaskId, T::TaskResult)
+    where
+        Self: Sync,
```

Comment:
Random Rust question: Why is "Self: Sync" needed here?  There is no self parameter to the function run_tasks.

Reply:
That's a good point. It's not needed here.